api/src/infrastructure/kubernetes/infrastructure.rs

```diff
@@ -0,0 +1,894 @@
+        let services = self.get_services_of_app(app_name).await?;
+        if services.is_empty() {
+            return Ok(services);
+        }
+
+        let futures = services
+            .iter()
+            .map(|service| self.stop_service(&app_name, &service))
+            .collect::<Vec<_>>();
+
+        for stop_service_result in join_all(futures).await {
+            trace!("stopped: {:?}", stop_service_result);
+            stop_service_result?;
+        }
+
+        Api::v1Namespace(self.client())
+            .delete(app_name, &DeleteParams::default())
+            .await?;
+
+        Ok(services)
+    }
+
+    async fn get_configs_of_app(&self, _app_name: &String) -> Result<Vec<ServiceConfig>, Error> {
+        let mut configs = Vec::new();
+        for service in self.get_services_of_app(_app_name).await? {
+            match service.container_type() {
+                ContainerType::ApplicationCompanion | ContainerType::ServiceCompanion => continue,
+                _ => {}
+            };
+
+            let image = service
+                .image()
+                .cloned()
+                .expect("Service should contain image");
+            let mut service_config = ServiceConfig::new(service.service_name().clone(), image);
+
+            if let Some(port) = service.port() {
+                service_config.set_port(port);
+            }
+
+            configs.push(service_config);
+        }
+
+        Ok(configs)
+    }
+
+    async fn get_logs(
+        &self,
+        app_name: &String,
+        service_name: &String,
+        from: &Option<DateTime<FixedOffset>>,
+        limit: usize,
+    ) -> Result<Option<Vec<(DateTime<FixedOffset>, String)>>, Error> {
+        let mut p = ListParams::default();
+        p.label_selector = Some(format!(
+            "{}={},{}={}",
+            APP_NAME_LABEL, app_name, SERVICE_NAME_LABEL, service_name,
+        ));
+        let pod = match Api::v1Pod(self.client())
+            .within(app_name)
+            .list(&p)
+            .await?
+            .into_iter()
+            .next()
+        {
+            Some(pod) => pod,
+            None => {
+                return Ok(None);
+            }
+        };
+
+        let mut p = LogParams::default();
+        p.timestamps = true;
+        p.since_seconds = from
+            .map(|from| {
+                from.timestamp()
+                    - pod
+                        .status
+                        .as_ref()
+                        .unwrap()
+                        .start_time
+                        .as_ref()
+                        .unwrap()
+                        .0
+                        .timestamp()
+            })
+            .filter(|since_seconds| since_seconds > &0);
+
+        let logs = Api::v1Pod(self.client())
+            .within(&app_name)
+            .log(&pod.metadata.name, &p)
+            .await?;
+
+        let logs = logs
+            .split("\n")
```

Comment:
For performance, it might be worthwhile to "parse" the response instead, by reading char-by-char and counting the newlines. For logs of more than one or two MB in size, this could be much faster than splitting into lines that will not be used later on. On the other hand, rust might generate views into the string here, rather than fresh strings, which could be fast enough.

Reply:
I created a new issue: #25